Optimise the namespace validation

idanov opened the pull request:
## Description

After adding validation for namespaces in https://github.com/kedro-org/kedro/pull/4603, Kedro was becoming prohibitively slow for any pipelines that had even slightly more complex pipelines that had many paths. The namespace validation was based on going through all paths, which could be exponentially many. This nicely explained issue has an example when the namespace validation just hangs https://github.com/kedro-org/kedro/issues/4943

I've added the example from the issue as a test.

## Development notes

I've reimplemented the algorithm to use the topologically sorted nodes and keep track of the visited pipeline at each node level. This way we go from `O(P)` down to `O(E + V)`, where `P` is the number of paths, `E` is the number of edges and `V` is the number of nodes. This is a very significant improvement, as `P` can be exponential in the number of nodes. When a violation of the namespace interruption occurs, then we use a `BFS` algorithm to find the shortest path starting from a node that exited the violated namespace. Bear in mind that there might be multiple paths that violate that, but only one will be reported (unlike in the previous algorithm), but that should be sufficient for debugging. 


## Developer Certificate of Origin
We need all contributions to comply with the [Developer Certificate of Origin (DCO)](https://developercertificate.org/). All commits must be signed off by including a `Signed-off-by` line in the commit message. [See our wiki for guidance](https://github.com/kedro-org/kedro/wiki/Guidelines-for-contributing-developers/).

If your PR is blocked due to unsigned commits, then you must follow the instructions under "Rebase the branch" on the GitHub Checks page for your PR. This will retroactively add the sign-off to all unsigned commits and allow the DCO check to pass.

## Checklist

- [x] Read the [contributing](https://github.com/kedro-org/kedro/blob/main/CONTRIBUTING.md) guidelines
- [x] Signed off each commit with a [Developer Certificate of Origin (DCO)](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/managing-repository-settings/managing-the-commit-signoff-policy-for-your-repository)
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [x] Added a description of this change in the [`RELEASE.md`](https://github.com/kedro-org/kedro/blob/main/RELEASE.md) file
- [x] Added tests to cover my changes
- [ ] Checked if this change will affect Kedro-Viz, and if so, communicated that with the Viz team

ankatiyar commented:
I'm getting two warnings when I try it out with interrupted namespaces:
```
[07/09/25 14:24:04] WARNING  /Users/ankita_katiyar/kedro/kedro/kedro/pipeline/pipeline.py warnings.py:110
                             :341: UserWarning: Namespace 'data_science' is interrupted                  
                             by nodes ['create_model_input_table_node'] and thus invalid.                
                               warn(                                                                     
                                                                                                         
                    WARNING  /Users/ankita_katiyar/kedro/kedro/kedro/pipeline/pipeline.py warnings.py:110
                             :341: UserWarning: Namespace 'data_science' is interrupted                  
                             by nodes ['create_model_input_table_node',                                  
                             'data_science.split_data_node'] and thus invalid.                           
                               warn(  
```
In my spaceflights project the `preprocess_companies` is namespaced `data_science` as well as the whole `data_science` pipeline has the same namespace. So the first warning is correct, but the second one isn't.

The algorithm largely makes sense!
I'm also surprised by how many times it gets triggered because it's in the `Pipeline` constructor which gets called several times, the warnings only show up once though.

idanov commented on kedro/pipeline/pipeline.py:
```diff
@@ -290,54 +290,69 @@ def __init__(  # noqa: PLR0913
+        return []  # pragma: no cover
+
+    def _validate_namespaces(self) -> None:
+        from warnings import warn
```
That's how it was in the previous code, can pull it out to the top.

idanov commented on kedro/pipeline/pipeline.py:
```diff
@@ -290,54 +290,69 @@ def __init__(  # noqa: PLR0913
+                last_namespace = parent.namespace or ""
+                curr_namespace = node.namespace or ""
+                # If curr_namespace was visited in paths leading to parent
+                if curr_namespace and curr_namespace in seen_namespaces[parent]:
```
That's a good point, will add it and it might make it more readable.

idanov commented on kedro/pipeline/pipeline.py:
```diff
@@ -290,54 +290,69 @@ def __init__(  # noqa: PLR0913
+                    )
+                # All visited namespaces for the current node get updated with the parent's visited namespaces
+                for ns, nodes in seen_namespaces[parent].items():
+                    visited[ns].update(nodes)
```
Not sure if that's the case, because these are all namespaces that were already visited. So technically, yes - in the worst case if all is added. That's why I was considering a persistent data structure, but in practice I doubt it will be such a bad time, e.g. 1000 nodes in one line, each in completely different namespaces (very unlikely scenario), that will result it about half a million iterations, which is definitely under 1 second execution time.

idanov replied:
Now when I think about it, adding only one node per namespace should be sufficient, as we only use the first one when reporting the path in case of interruption. Let's leave these comments in the PR for posterity in case we see slowdowns here in the future, I don't want to apply this optimisation because I don't want to risk correctness (as I haven't thought out very well if this might have implication on the correctness of the solution at the end).

idanov commented:
> I'm getting two warnings when I try it out with interrupted namespaces:
> 
> ```
> [07/09/25 14:24:04] WARNING  /Users/ankita_katiyar/kedro/kedro/kedro/pipeline/pipeline.py warnings.py:110
>                              :341: UserWarning: Namespace 'data_science' is interrupted                  
>                              by nodes ['create_model_input_table_node'] and thus invalid.                
>                                warn(                                                                     
>                                                                                                          
>                     WARNING  /Users/ankita_katiyar/kedro/kedro/kedro/pipeline/pipeline.py warnings.py:110
>                              :341: UserWarning: Namespace 'data_science' is interrupted                  
>                              by nodes ['create_model_input_table_node',                                  
>                              'data_science.split_data_node'] and thus invalid.                           
>                                warn(  
> ```
> 
> In my spaceflights project the `preprocess_companies` is namespaced `data_science` as well as the whole `data_science` pipeline has the same namespace. So the first warning is correct, but the second one isn't.
> 
> The algorithm largely makes sense! I'm also surprised by how many times it gets triggered because it's in the `Pipeline` constructor which gets called several times, the warnings only show up once though.

Interesting, could you please share your pipeline, as it is quite difficult for me to reproduce your pipeline...

idanov commented on kedro/pipeline/pipeline.py:
```diff
@@ -290,54 +291,71 @@ def __init__(  # noqa: PLR0913
         self._toposorted_nodes: list[Node] = []
         self._toposorted_groups: list[list[Node]] = []
-        self._validate_namespaces(node_parents)
+        self._validate_namespaces()
```
That won't make a huge difference, but I guess we can add it.

idanov commented:
I have fixed the issue discovered by @ankatiyar and changed the behaviour to report only the first interrupted namespace level (as the previous behaviour was potentially producing too many warnings). I've corrected one of the tests and added a new property to the `Node` class, as this is a property that @rashidakanchwala might find useful in Kedro Viz (as there's currently already an implementation of this called `exlode_namespaces`).

merelcht reviewed (approved):
Tested this against @ankatiyar's example, and it's working as expected 👍 Thanks @idanov ! ⭐

ankatiyar reviewed (approved):
Manually tested with different cases, works perfectly! 💯
